### DISCUSSION TITLE
Why is browser compatibility a thing?

### DISCUSSION THREAD

**Discussion by ghost** [denoland/deno#8962] [Q&A] [answered]
Hi everyone,
why is browser compatibility for Deno a thing?
Shouldn't it be as fast as possible on servers?

**Comment by mvolkmann** [comment, marked as answer]
I think the point is to have a compatible API so the the knowledge developers already have about browser APIs carries over. Examples include the Fetch API and the `console` methods.

**Reply by ghost** [reply]
So Browser support means to use the browser specific functions and not to run deno inside a browser?

**Reply by mvolkmann** [reply]
That is my understanding.

**Comment by ghost** [comment]
> why is browser compatibility for Deno a thing?

Browser compatibility allows one to write the same code and use it in multiple places, e.g. on Firefox, Safari, Deno, etc, without (major) change to the source code.

*Most* of the browser APIs are powerful, useful features that developers have wanted or needed, they also have very well defined algorithms and behavior, making (some of) them easy to port anywhere.

But, there seems to be a misunderstanding here,
> Shouldn't it be as fast as possible on servers?

Performance and browser compatibility are **completely** unrelated.
The only things that matter regarding performance are the implementation, and how *you* put the API to use.

There is also more to performance than just "speed," ex: memory consumption.

Take the `fetch` API, it returns a `Promise<Response>`, `Response`s can be used as the stream is being received.
Here's an example:
```ts
const { body } = await fetch(...);

for await ( const array of body ) {
    ...
}
```
using this stream of u8 arrays, you could do something like assigning the array to a global variable, preventing it from being garbage collection, forcing the runtime to allocate more memory, ultimately slowing the code down. If you do that, then it's partially your fault.

But, let's say that Deno wanted to implement a way to send a request, and receive data, what if it read everything in sync, and only yielded once the entire response was read?

In that case, you would complain that Deno's API isn't good, but now that situation won't occur at all, as they simply implement the standardized API. As a bonus, any bugs can quickly be spotted by those who are familiar with a browser implementation.

~~But if Deno is slow, it's because it is built on V8, instead of JSC~~
